File: 234. Palindrome Linked List/solution_3.java
//use stack only
//7.58%
//from discuss
//not a great method
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null)  return true;
        Stack<Integer> stack = new Stack<Integer>();
        ListNode ptr = head;
        while (ptr != null) {
            stack.push(ptr.val);
            ptr = ptr.next;
        }
        while (head != null) {
            if (head.val != stack.pop()) {
                return false;
            }
            head = head.next;
        }
        return true;
    }
}

File: 148. Sort List/solution_3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode *merge(ListNode *l, ListNode *r) {
        if (!l) return r;
        if (!r) return l;
        ListNode *pre = new ListNode(-1), *ptr = pre;
        while (l && r) {
            if (l->val < r->val) {
                ptr->next = l;
                l = l->next;
            }
            else {
                ptr->next = r;
                r = r->next;
            }
            ptr = ptr->next;
        }
        if (l)  ptr->next = l;
        if (r)  ptr->next = r;
        return pre->next;
    }
public:
    ListNode* sortList(ListNode* head) {
        if (!head)  return nullptr;
        if (!head->next)    return head;
        ListNode *fast = head->next, *slow = head;
        while (fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }
        ListNode *mid = slow->next;
        slow->next = nullptr;
        head = sortList(head);
        mid = sortList(mid);
        head = merge(head, mid);
        return head;
    }
};